---
title: "How to run the Bioinformatics Scripts"
output: 
  word_document:
  html_document: 
    toc: yes
vignette: >
  %\VignetteIndexEntry{qiime2}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval=FALSE)
library(aquaman)
library(kraken)
library(ggplot2)
library(readr)
```

### Summary

-   Download data from operator or SEPA lab

-   Run `Qiime2` based taxonomic assignment script

-   Assess quality, export and format the taxa data

-   Calculate the IQI score based on the bacteria taxa

-   Upload the bacteria based IQI scores to the Access database

-   Calculate the Area footprint in the Spotfire area calculator tool

-   Report results to Compliance team

### Installation

To run the [qiime2](https://qiime2.org/) based taxa assignment, [install](https://docs.qiime2.org/2022.8/install/) qiime2 and conda software libraries. These run on MacOS or Linux. For Windows users, recommend using Windows Subsystem for Linux (WSL).

To run the 'IQI meiometbar SEPA screening tool' model (not sure what it should actually be called) requires [R](https://www.r-project.org/). Install R on Windows virtual machine, and [Rstudio](https://posit.co/download/rstudio-desktop/), and then install R library [aquaman](https://github.com/aquaMetrics/aquaman). This makes it easier to run the IQI model and incorporates the exact same model produced by SAMS.

### Saving files

The Operator or SEPA lab will provide a link to where sequenced data is stored.

1.  Check these files are .fastq.gz format.
2.  Save the files tosharepoint Ecology drive '[results](https://scottishepa.sharepoint.com/:f:/r/sites/Ecology/DNA/Results?csf=1&web=1&e=806wgL)' folder.
3.  Name the folder with the unique survey name.

### Run Assignment

1.  On Windows, open WSL ubuntu 20.04, and 'mount' (mnt), the Windows folder containing the sequence data files (sync-ed folder your one drive first). For example:

    ```{bash}

    cd /mnt/c/Users/Tim.Foster/OneDrive\ -\ Scottish\ Environment\ Protection\ Agency/DNA/Results/ 
    ```

    NOTE: For reference the \`qiime2-taxa-assign.txt\` file can be found here:

    [aquaman/inst/extdat/qiime2-bash-script at main · aquaMetrics/aquaman (github.com)](https://github.com/aquaMetrics/aquaman/tree/main/inst/extdat/qiime2-bash-script)

2.  Activate conda python environment required for Qiime2.

    ```{bash}
    conda activate qiime2-2022.2
    ```

3.  Check your VM has at least 10GB free storage space. And close all background programs (e.g. Teams, Web browser, Email etc)

4.  Run the bash script and provide a relative path to sequenced files.

    ```{bash}
    bash qiime2-taxa-assign.txt directory/containing/files
    ```

5.  This produces a number of output files which will be stored in the sub-directory with dna-files.

    NOTE: If changes made to `qiime2-taxa-assign.txt` file, then may require cleaning the file of carriage returns introduced by editing.

    `tr -d "/r" < qiime2-taxa–assign.txt > qiime2-taxa-assign-cleaned.txt`

NOTE: If the bash script has run and output files exist, then the script checks if an output file exists and only runs the section of code if there is no output file.

### Quality Control

Refer to the [DNA SOP](https://www.sepa.org.uk/media/594636/guidance-on-the-use-of-dna-in-monitoring-compliance-with-seabed-mixing-zone-area-limits.pdf) for current requirements on number of reads required. Assess the '...demuxed_paired_end.qzv' files using [Qiime2 Viewer](https://view.qiime2.org/) where the number of merged reads can be verified.

### Export Taxa

Taxa information must be exported to .CSV for later processing in R.

1.  Drag and drop filenames containing '...S16....qzv' to [Qiime2 Viewer](https://view.qiime2.org/).
2.  Visually check samples diverse samples and no outliers.
3.  Select Taxonomic Level 7 from drop-down and click the 'Download as CSV' button.
4.  Save .CSV files along side the .qzv files in the relevant dna-files/ directory.

### Format Assigned Taxa

Prior to running the IQI assessment of the taxonomic data, the .CSV data is converted to the correct input format.

1.  Open RStudio

Copy the following commands - adjusting to point to your directory of your taxa.csv file.

```{r, eval=FALSE}

library(aquaman)
taxa <- aquaman:::import_qiime2(path = "C:\\Users\\Tim.Foster\\OneDrive - Scottish Environment Protection Agency\\DNA\\Results\\BaseCalls\\taxa.csv")


```

### Calculate IQI

Use the `iqi()` function to run the IQI model and calculate IQI. The output from `iqi()` function is reproducible. The same input will provide the same output. The meiometbar scripts (ScopeTool - BioSS) re-trains the IQI from scratch each time, this takes longer and is not reproducible. I think due to `set.seed()` not being used 'fix' the randomness used when training the RF model. In other words, running the `RRF()` function with the same input data will generate a slightly different model each time. Setting `set.seed()` to the same value before running `RRF()` will create the same model each time.

```{r, eval=FALSE}
iqi_score <- aquaman::iqi(taxa)
iqi <- data.frame("sample" = names(iqi_score), "iqi" = iqi_score)
```

Examples of where to save:

```{r, eval=FALSE}

write.csv(iqi, "C:/Users/Tim.Foster/OneDrive - Scottish Environment Protection Agency/DNA/Results/aquaculture-operator-data/output/iqi-kingairloch-update.csv", row.names = FALSE)


```

Example of quick plot of IQI scores (not really required - this is example of test data for against external operators IQI scores)

```{r}

iqi_compare <- read_csv("C:/Users/Tim.Foster/OneDrive - Scottish Environment Protection Agency/DNA/Results/aquaculture-operator-data/output/iqi-kingairloch-update.csv")


# rename long variables to make plot better fit
# iqi_compare <- dplyr::rename(iqi_compare,
#                     "all_sepa_mowi_rf" =  all_sepa_mowi_rf_test_predictions,
#                     "sepa_screening_iqi" =  sepa_iqi_aquaman_update
#                    )

# pivot data pre-plotting
iqi_pivot <- tidyr::pivot_longer(iqi_compare, cols =c(
  "MOWI_IQI",
  "SEPA_IQI",
  "INFAUNA_IQI"),
  names_to = "IQI type", 
  values_to = "IQI"
)

# plot
plot <- ggplot(iqi_pivot, 
               aes(distance, IQI, group=`IQI type`)) +
  geom_line(aes(colour=`IQI type`)) + geom_point(aes(colour=`IQI type`)) +
    geom_hline(yintercept=c(rep(0.64,63))) +
  facet_wrap(facets = as.factor(iqi_pivot$transect)) + ggtitle("Kingairloch") 

plot 
```

### Format for Access

Outputs may need to be format to fit with existing internal Access database. Copy output into Access database:

-   C:\\Users\\Tim.Foster\\OneDrive - Scottish Environment Protection Agency\\FishFarmDatabaseMaster\\MPFFSelfMonDatabase.accdb

    This requires Access software limited to some virtual desktop machines.

### Run Spotfire Area Tool

Calculate area in Spotfire tool (maybe need data refresh overnight to get latest data from Access)?

### Compliance

Submit results to Compliance team via email to complete compliance review.
